File: web/assets/js/semantica.js
/* =========================================================================
   logica/semantica.py → JavaScript
   Evaluación semántica: tablas de verdad y clasificación
   (Tautología / Contradicción / Contingencia).
   ========================================================================= */
"use strict";

(function () {
  var LP = window.LogicaParser;
  var S = {};

  S.TAUTOLOGIA = "Tautología";
  S.CONTRADICCION = "Contradicción";
  S.CONTINGENCIA = "Contingencia";
  S.MAX_ATOMOS_TABLA = 5;

  function evaluar(n, asg) {
    if (n instanceof LP.AtomNode) return !!asg[n.nombre];
    if (n instanceof LP.NegNode) return !evaluar(n.hijo, asg);
    var izq = evaluar(n.izq, asg);
    var der = evaluar(n.der, asg);
    switch (n.op) {
      case "∧": return izq && der;
      case "∨": return izq || der;
      case "→": return (!izq) || der; // el condicional solo falla con V→F
      case "↔": return izq === der;
    }
    throw new Error("Operador desconocido: " + n.op);
  }

  // Combinaciones de valores: mismo orden que itertools.product (V...V → F...F).
  function combinaciones(atoms) {
    var filas = [];
    var total = Math.pow(2, atoms.length);
    for (var mask = 0; mask < total; mask++) {
      var val = total - 1 - mask;
      var asg = {};
      for (var j = 0; j < atoms.length; j++) {
        asg[atoms[j]] = !!((val >> (atoms.length - 1 - j)) & 1);
      }
      filas.push(asg);
    }
    return filas;
  }

  function tablaVerdad(n) {
    var vars_ = LP.atomos(n);
    var nv = vars_.length;
    var datos = { atoms: vars_, filas: [], valores: [], excede_limite: nv > S.MAX_ATOMOS_TABLA };
    if (datos.excede_limite) return datos;
    combinaciones(vars_).forEach(function (asg) {
      datos.filas.push(asg);
      datos.valores.push(evaluar(n, asg));
    });
    return datos;
  }

  function subexpresiones(n) {
    var resultado = [n];
    LP.hijos(n).forEach(function (h) { resultado = resultado.concat(subexpresiones(h)); });
    return resultado;
  }

  function tablaVerdadCompleta(n) {
    var vars_ = LP.atomos(n);
    var subnodos = subexpresiones(n);
    var columnas = [];
    subnodos.forEach(function (sub) {
      var t = LP.aCadena(sub);
      if (columnas.indexOf(t) === -1) columnas.push(t);
    });
    var total = vars_.length;
    if (total > S.MAX_ATOMOS_TABLA) return { columnas: columnas, filas: [], excede_limite: true };

    var filas = [];
    combinaciones(vars_).forEach(function (asg) {
      var fila = {};
      subnodos.forEach(function (sub) {
        var t = LP.aCadena(sub);
        fila[t] = evaluar(sub, asg) ? "V" : "F";
      });
      filas.push(fila);
    });
    return { columnas: columnas, filas: filas, excede_limite: false };
  }

  function clasificar(n) {
    var tabla = tablaVerdad(n);
    var valores = tabla.valores;
    var canon = LP.aCadena(n);

    if (tabla.excede_limite) {
      return {
        canonical: canon,
        clase: "No determinable" + " (>" + S.MAX_ATOMOS_TABLA + " átomos: " + tabla.atoms.length + ")",
        verdaderos: null,
        falsos: null,
      };
    }

    var verdaderos = 0;
    valores.forEach(function (v) { if (v) verdaderos++; });
    var falsos = valores.length - verdaderos;
    var clase;
    if (falsos === 0) clase = S.TAUTOLOGIA;
    else if (verdaderos === 0) clase = S.CONTRADICCION;
    else clase = S.CONTINGENCIA;

    return { canonical: canon, clase: clase, verdaderos: verdaderos, falsos: falsos };
  }

  function tablaHtml(datos) {
    var columnas = datos.columnas;
    var filas = datos.filas || [];
    if (datos.excede_limite) {
      return '<p class="aviso">La tabla tendría 2^' + columnas.length +
             " filas; se omite por superar el límite de trabajabilidad.</p>";
    }
    var th = columnas.map(function (c) { return "<th>" + LP.escapar(c) + "</th>"; }).join("");
    var cuerpo = filas.map(function (fila) {
      var tds = columnas.map(function (c) { return "<td>" + LP.escapar(fila[c]) + "</td>"; }).join("");
      return "<tr>" + tds + "</tr>";
    }).join("");
    return '<table class="tabla-verdad"><thead><tr>' + th + "</tr></thead><tbody>" + cuerpo + "</tbody></table>";
  }

  S.evaluar = evaluar;
  S.tablaVerdad = tablaVerdad;
  S.tablaVerdadCompleta = tablaVerdadCompleta;
  S.clasificar = clasificar;
  S.tablaHtml = tablaHtml;

  window.Semantica = S;
})();

File: logica/semantica.py
# -*- coding: utf-8 -*-
"""Módulo de semántica: evaluación con tablas de verdad.

Define cómo asignar valores de verdad a cada proposición atómica
y cómo clasificar una FBF en **Tautología**, **Contradicción** o **Contingencia**.
"""

from __future__ import annotations

import html
from itertools import product
from typing import Dict, List

from .parser import AtomNode, NegNode, BinNode, a_cadena, atomos, hijos

# Clasificaciones posibles de una FBF según su tabla de verdad.
TAUTOLOGIA = "Tautología"
CONTRADICCION = "Contradicción"
CONTINGENCIA = "Contingencia"

# Límite de átomos para construir la tabla de verdad (2^n filas).
MAX_ATOMOS_TABLA = 5


def evaluar(nodo, asignacion: Dict[str, bool]) -> bool:
    """Evalúa el AST con una asignación de valores (dict átomo -> bool).

    Tablas de verdad de operadores:
        ¬A  : negación (invierte).
        A∧B : verdadero solo si ambos lo son.
        A∨B : verdadero si al menos uno lo es.
        A→B : falso solo si A es V y B es F.
        A↔B : verdadero si ambos coinciden.
    """
    if isinstance(nodo, AtomNode):
        return bool(asignacion[nodo.nombre])
    if isinstance(nodo, NegNode):
        return not evaluar(nodo.hijo, asignacion)
    izq = evaluar(nodo.izq, asignacion)
    der = evaluar(nodo.der, asignacion)
    if nodo.op == "∧":
        return izq and der
    if nodo.op == "∨":
        return izq or der
    if nodo.op == "→":
        return (not izq) or der          # el condicional solo falla con V→F
    if nodo.op == "↔":
        return izq == der
    raise ValueError(f"Operador desconocido: {nodo.op!r}")


def _combinaciones(atoms: List[str]):
    """Genera todas las combinaciones de valores de verdad (producto cartesiano)."""
    return product([True, False], repeat=len(atoms))


def tabla_verdad(nodo) -> dict:
    """Construye la tabla de verdad de la fórmula.

    Returns:
        dict con ``atoms`` (columnas), ``filas`` (asignaciones) y ``valores``
        (resultado de la fórmula en cada fila). Si hay demasiados átomos se
        indica el límite.
    """
    vars_ = atomos(nodo)
    n = len(vars_)
    datos = {"atoms": vars_, "filas": [], "valores": [], "excede_limite": n > MAX_ATOMOS_TABLA}
    if datos["excede_limite"]:
        return datos
    for combinacion in _combinaciones(vars_):
        asg = dict(zip(vars_, combinacion))
        datos["filas"].append(asg)
        datos["valores"].append(evaluar(nodo, asg))
    return datos


def _subexpresiones(nodo) -> List:
    """Lista de todas las sub-fórmulas del árbol en orden pre-orden (raíz primero)."""
    resultado = [nodo]
    for sub in hijos(nodo):
        resultado.extend(_subexpresiones(sub))
    return resultado


def tabla_verdad_completa(nodo) -> dict:
    """Tabla de verdad enriquecida: una columna por cada sub-fórmula.

    Devuelve ``columnas`` (encabezados) y ``filas`` (valores como texto V/F).
    """
    vars_ = atomos(nodo)
    subnodos = _subexpresiones(nodo)
    # Solo conservamos una columna por cada subfórmula textual distinta.
    columnas: List[str] = []
    for sub in subnodos:
        texto = a_cadena(sub)
        if texto not in columnas:
            columnas.append(texto)

    total = len(vars_)
    filas = []
    if total > MAX_ATOMOS_TABLA:
        return {"columnas": columnas, "filas": [], "excede_limite": True}

    for combinacion in _combinaciones(vars_):
        asg = dict(zip(vars_, combinacion))
        fila = {}
        for sub in subnodos:
            texto = a_cadena(sub)
            valor = evaluar(sub, asg)
            fila[texto] = "V" if valor else "F"
        filas.append(fila)
    return {"columnas": columnas, "filas": filas, "excede_limite": False}


def clasificar(nodo) -> dict:
    """Clasifica la FBF según su tabla de verdad.

    Returns:
        dict con la clase, el conteo de valores verdaderos/falsos y
        la tabla asociada.
    """
    tabla = tabla_verdad(nodo)
    valores = tabla["valores"]

    canon = a_cadena(nodo)
    if tabla.get("excede_limite"):
        # No podemos probar todas las filas: reportamos indeterminado.
        return {
            "canonical": canon,
            "clase": "No determinable"
            f" (>{MAX_ATOMOS_TABLA} átomos: {len(tabla['atoms'])})",
            "verdaderos": None,
            "falsos": None,
            "tabla": tabla,
        }

    verdaderos = sum(1 for v in valores if v)
    falsos = len(valores) - verdaderos
    if falsos == 0:
        clase = TAUTOLOGIA
    elif verdaderos == 0:
        clase = CONTRADICCION
    else:
        clase = CONTINGENCIA

    return {
        "canonical": canon,
        "clase": clase,
        "verdaderos": verdaderos,
        "falsos": falsos,
        "tabla": tabla,
    }


def tabla_html(datos: dict) -> str:
    """Renderiza la tabla de verdad como HTML (para insertar en la GUI)."""
    columnas = datos["columnas"]
    filas = datos.get("filas", [])
    if datos.get("excede_limite"):
        return (
            '<p class="aviso">La tabla tendría '
            f"2^{len(columnas)} filas; "
            "se omite por superar el límite de trabajabilidad.</p>"
        )

    encabezados = "".join(f"<th>{html.escape(c)}</th>" for c in columnas)
    cuerpo = []
    for fila in filas:
        celdas = "".join(f"<td>{html.escape(fila[c])}</td>" for c in columnas)
        cuerpo.append(f"<tr>{celdas}</tr>")
    return (
        f'<table class="tabla-verdad"><thead><tr>{encabezados}</tr></thead>'
        f"<tbody>{''.join(cuerpo)}</tbody></table>"
    )